Migrate `JobFilesAPIController` to FastAPI (excluding TUS uploads)

domgz opened the pull request:
As part of the development of an integration of Galaxy with [ARC (Advanced Resource Connector)](https://www.nordugrid.org/arc/arc7/index.html) as a Pulsar job runner, tweaks to the `JobFilesAPIController` are needed. None of such tweaks are included in this PR, but it makes sense to implement them building upon a FastAPI endpoint rather than a legacy WSGI one; and the first step to do that is migrating the controller to FastAPI.

`FastAPIJobFiles` is the new, FastAPI version of `JobFilesAPIController`. The endpoints that have been migrated should exhibit exactly the same behavior as the old ones from `FastAPIJobFiles`.

Endpoints dedicated to TUS uploads work in tandem with the WSGI middleware `TusMiddleware` from the `tuswsgi` package. WSGI middlewares and endpoints are injected into the FastAPI app after FastAPI routes as a single sub-application `wsgi_handler` using `app.mount("/", wsgi_handler)`, meaning that requests are passed to the `wsgi_handler` sub-application (and thus to `TusMiddleware`) only if there was no FastAPI endpoint defined to handle them. Therefore, they cannot be migrated to FastAPI unless `TusMiddleware` is also migrated to ASGI. I am postponing that migration, because the ARC integration needs to be delivered soonish.

I also included three new tests for existing functionality: writing from uploads done with the nginx_upload_module, from TUS uploads and using the parameter `__file`.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Migrate `JobFilesAPIController` to FastAPI (excluding TUS uploads)

`FastAPIJobFiles` is the new, FastAPI version of `JobFilesAPIController`. The endpoints that have been migrated should exhibit exactly the same behavior as the old ones from `FastAPIJobFiles`. Something to keep in mind is that while FastAPI has some extra built-in features that the legacy WSGI system did not have, such as answering HEAD requests, those do not work because of the way legacy WSGI endpoints are injected into the FastAPI app (using `app.mount("/", wsgi_handler)`), meaning that for example, HEAD requests are passed to the `wsgi_handler` sub-application.

Endpoints dedicated to TUS uploads work in tandem with the WSGI middleware `TusMiddleware` from the `tuswsgi` package. As explained above, WSGI middlewares and endpoints are injected into the FastAPI app after FastAPI routes as a single sub-application `wsgi_handler` using `app.mount("/", wsgi_handler)`, meaning that requests are passed to the `wsgi_handler` sub-application (and thus to `TusMiddleware`) only if there was no FastAPI endpoint defined to handle them. Therefore, they cannot be migrated to FastAPI unless `TusMiddleware` is also migrated to ASGI.

domgz committed:
Generate unique operation ids for `/api/jobs/{job_id}/files`

Work around a bug in FastAPI (https://github.com/fastapi/fastapi/issues/13175) that assigns the same operation id to both request methods GET and HEAD of the endpoint `/api/jobs/{job_id}/files` when using the `@router.api_route()` decorator with `methods=["GET", "HEAD"]` as keyword argument.

domgz committed:
Accept `path` and `job_key` both as query and form parameters for POST requests to `/api/jobs/{job_id}/files`

Pulsar formats the `path` and `job_key` parameters as query parameters when submitting POST requests to `/api/jobs/{job_id}/files`. However, many Galaxy tests format them as form parameters. The only way to keep the endpoint working as it should (as it worked before the migration to FastAPI) is to accept both query and form parameters.

domgz committed:
Fix OpenAPI docs for `path` and `job_key` as form parameters for POST requests to `/api/jobs/{job_id}/files`

FastAPI will not use the parameter aliases of form parameters in the OpenAPI docs, but the name of their Python variables. Therefore, the API docs show `path_form` and `job_key_form`. Rename them so that the API docs show the correct parameter names.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -1,67 +1,173 @@
 
-        :rtype:     binary
-        :returns:   contents of file
+    # FastAPI answers HEAD requests automatically for GET endpoints. However, because of the way legacy WSGI endpoints
```
Where's that info coming from ? This has been discussed but was never implemented (last attempt https://github.com/fastapi/fastapi/pull/13509).

domgz replied:
If you return `FileResponse`, then it's true https://github.com/fastapi/fastapi/issues/1773#issuecomment-662795361. I **incorrectly assumed it was generally true**, because I just tried it with this endpoint.

I changed the comments so that they are precise 8def3a7. Thanks for letting me know about this.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -1,67 +1,173 @@
+        *(_args := ["/api/jobs/{job_id}/files"]),
+        **(
+            _kwargs := dict(
+                summary="Get a file required to staging a job.",
+                responses={
+                    200: {
+                        "description": "Contents of file.",
+                        "content": {"application/json": None, "application/octet-stream": {"example": None}},
+                    },
+                    400: {
+                        "description": (
+                            "File not found, path does not refer to a file, or input dataset(s) for job have been purged."
+                        )
+                    },
+                },
+            )
+        ),
+    )
+    @router.head(*_args, **_kwargs)  # type: ignore[name-defined]
+    # remove `@router.head(...)` when ALL endpoints have been migrated to FastAPI
```
Can you follow the ordinary way of setting up fastapi endpoints, i.e. no argument unpacking and I don't think the explicit responses and summary keyword are needed either.

domgz replied:
This is the result of adding the keywords. 

<img width="1719" height="65" alt="grafik" src="https://github.com/user-attachments/assets/4988d490-db11-4fa5-a72f-39b4eb34385b" />

...

<img width="1715" height="804" alt="grafik" src="https://github.com/user-attachments/assets/44df5f6e-fa58-4368-aa25-2cbfb350452e" />

I don't think there harm in neatly documenting the endpoint, especially because the cause of HTTP 400 errors is hard to guess.

Regarding argument unpacking, refer to [this answer](https://github.com/galaxyproject/galaxy/pull/20235#discussion_r2339584898).

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -1,67 +1,173 @@
+router = Router(
+    # keep the endpoint in the undocumented section of the API docs `/api/docs`, as all endpoints from `FastAPIJobFiles`
+    # are certainly not meant to represent part of Galaxy's stable, user facing API
+    tags=["undocumented"]
```
That's meant for the stuff that hasn't been ported. There's no harm documenting these and making them available for client libraries as long as these are marked as internal.

domgz replied:
Assigned "jobs" tag (c5c7d34).

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -1,67 +1,173 @@
+    #
+    @router.get(
+        # simplify me (remove `_args` and `_kwargs` defined using the walrus operator) when ALL endpoints have been
+        # migrated to FastAPI, this is a workaround for FastAPI bug https://github.com/fastapi/fastapi/issues/13175
```
That's maybe a little excessive, you can also pass in an explicit operation_id if you did this to avoid the warning.

domgz replied:
The point of having both decorators `@router.get()` and `@router.head()` is indeed avoiding the warning.

If I have only one decorator `@router.api_route()`, then passing an explicit `operation_id` does not solve anything (both GET and HEAD get the same operation id).

If I use both decorators, then I never have to pass an explicit `operation_id`, because FastAPI will generate two different ones for me.

Regarding argument unpacking, the idea is that the same method `index()` has to be reused both for GET and HEAD. I am forced to use the two-decorator approach (to avoid the warning) and by definition, both the GET and HEAD endpoints must be defined in exactly the same way. I use argument unpacking to avoid repeating the arguments of the first decorator in the second decorator.

I realize however, that including both GET and HEAD in the API schema is redundant. Thus, I added `include_in_schema=False` to the `@router.head()` decorator (bed78be).

mvdbeek replied:
Sorry, I don't think my comment is addressed. Please prefer a bit of repetitive code over disabling type checks.

domgz replied:
Since I set `include_in_schema=False`, `@router.head()` can now skip the kwargs and there would then be little repetition (only the path). Seems ok if it's just that what's repeated.

domgz replied:
Addressed in b53fa359.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -1,67 +1,173 @@
+
+    This method does not force the client to provide the parameter, it could simply not submit the parameter in either
+    format. To force the client to provide the parameter, coerce the output of the method to a string, i.e.
+    `job_key: str = Depends(job_key_query_or_form)` so that FastAPI responds with status code 500 when the parameter is
```
That should be a validation error, not an internal server error.

domgz replied:
That's correct, fixed in [62d3499](https://github.com/galaxyproject/galaxy/pull/20235/commits/62d3499171dd4411c5cc33891bd012ec795cec51)·

domgz committed:
Fix HTTP error code typo

Change docstrings of `job_key_query_or_form()` and `path_query_or_form()` to reference the correct HTTP error code (400).

domgz committed:
Merge branch 'dev' into job_files_fastapi_migration

domgz committed:
Cleanup changes using pyupgrade

There was a small merge conflict with 7f89e0e, make sure changes are compliant with said commit.

mvdbeek reviewed (commented):
here's the (incomplete) review i never submitted, maybe you can take a look at this and I'll come back to this next week

domgz committed:
Hide HEAD method for `/api/jobs/{job_id}/files` in API docs

Including both GET and HEAD in the API schema for the same endpoint is redundant.

domgz committed:
Correct comment claiming that FastAPI answers HEAD requests automatically

It is only true for endpoints returning `FileResponse`, see https://github.com/fastapi/fastapi/issues/1773#issuecomment-662795361.

domgz committed:
Assign "jobs" tag to `FastAPIJobFiles` router

Group job files API endpoints together with jobs API endpoints.

mvdbeek reviewed (commented):
This is a pretty important and performance sensitive endpoint. Have you tested all of the variants? Can you run a performance comparison (for speed and memory consumption) between the old and new endpoints ?

domgz commented on lib/galaxy/webapps/galaxy/api/job_files.py:
```diff
@@ -127,76 +240,41 @@ def create(self, trans, job_id, payload, **kwargs):
-        tus_patch.
-        """
-        pass
+                with open(path, "wb") as destination:
+                    shutil.copyfileobj(cast(IO[bytes], input_file), destination)
```
This is the potential performance mentioned [in this comment](https://github.com/galaxyproject/galaxy/pull/20235#issuecomment-3319028191). For readability I just copied the new code below (the comment above the code explains the problem).

```python
            # prior to migrating to FastAPI, this operation was done more efficiently for all cases using
            # `shutil.move(input_file_path, path)`, but FastAPI stores the uploaded file as
            # `tempfile.SpooledTemporaryFile`
            # (https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile), so now there is not even
            # a path where uploaded files can be accessed on disk
            if input_file_path:
                shutil.move(input_file_path, path)
            else:
                with open(path, "wb") as destination:
                    shutil.copyfileobj(cast(IO[bytes], input_file), destination)
```

It's tricky because the limitation is in FastAPI's design.

mvdbeek replied:
you should be able to call rollover on the file and move it then ?

domgz replied:
Thanksss, that's just what I was looking for!

domgz replied:
Hmm there is a problem with this. Besides storing the data in memory until the file is rolled over to disk, `tempfile.SpooledTemporaryFile` operates exactly as [`tempfile.TemporaryFile`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile). So even if [`rollover()`](https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile.rollover) is called, **there is no path for that file** (and there is nothing to call `shutil.move()` on).

> Under Unix, the directory entry for the file is either not created at all or is removed immediately after the file is created. Other platforms do not support this; your code should not rely on a temporary file created using this function having or not having a visible name in the file system.

Definitely writing the file to disk is twice suboptiomal... :thinking:

domgz committed:
Declare `@router.head()` explicitly rather than via argument unpacking

The point of having both decorators `@router.get()` and `@router.head()` is working around FastAPI issue https://github.com/fastapi/fastapi/issues/13175.

If there is only one decorator `@router.api_route()`, then both GET and HEAD get the same operation id. If both decorators are used, then FastAPI generates two different operation ids.

The idea behind argument unpacking is that the same method `index()` has to be reused both for GET and HEAD. The two-decorator approach is needed to avoid the duplicate operation id warning and by definition, both the GET and HEAD endpoints must be defined in exactly the same way. Argument unpacking allows to avoid repeating the arguments of the first decorator in the second decorator.

However, including both GET and HEAD in the API schema is redundant. Commit bed78be adds `include_in_schema=False` to the `@router.head()`.

That implies that there is now not so much benefit in the argument unpacking trick, because the kwargs are no longer needed and there would then be little repetition (only the path). Thus, declare `@router.head()` explicitly.
